### DISCUSSION TITLE
Mesh attributes for types with memory allocation

### DISCUSSION THREAD

**Discussion by sebmestrallet** [BrunoLevy/geogram#130] [Q&A] [answered]
I saw in the [wiki of mesh attributes](https://github.com/BrunoLevy/geogram/wiki/Mesh#attributes):

> _advanced_ In addition, arbitrary types can be used (including classes that do memory allocation). Before using a new attribute type, one needs to declare it (only once, at initialization), using `geo_register_attribute_type<type>("name")` (see [geogram/common.cpp](https://github.com/BrunoLevy/geogram/blob/main/src/lib/geogram/basic/common.cpp) for example).

So I tried to add :
```cpp
// ...
#include <geogram/basic/memory.h>
// ...
geo_register_attribute_type<GEO::vector<index_t>>("indices_vector");
// ...
```
to `common.cpp`, but I got:
```txt
...
geogram/basic/string.h:334:24: error: no match for ‘operator>>’ (operand types are ‘std::istringstream’ {aka ‘std::__cxx11::basic_istringstream<char>’} and ‘GEO::vector<unsigned int>’)
  334 |             return (in >> value) && (in.eof() || ((in >> std::ws) && in.eof()));
      |                    ~~~~^~~~~~~~~
...
geogram/basic/string.h:244:17: error: no match for ‘operator<<’ (operand types are ‘std::ostringstream’ {aka ‘std::__cxx11::basic_ostringstream<char>’} and ‘const GEO::vector<unsigned int>’)
  244 |             out << value;
      |             ~~~~^~~~~~~~
...
```

Are attributes for custom types with memory allocation still supported?

Is the compilation error related to file I/O? If so, what are the specifications for serialization of custom type attributes?

**Comment by BrunoLevy** [comment, marked as answer]
Now it is supported:

Update: implemented the mechanism for handling attribute types with ctor/dtor/copy ctor/operator= 
(Now you can use attributes of type `std::string` for instance, just make sure you call:
```
geo_register_attribute_type<std::string>("string");
``` 
before using them. You'll also need to have an `operator >>` and `operator <<` for your type.

An example is [here](https://github.com/BrunoLevy/geogram/blob/main/src/tests/test_lifecycled_attribute/main.cpp)

~Hello, I confirm that attributes for types that are not POD (Plain Ordinary Datatype) is *not* supported.~

~It used to be supported in the past, but it introduces much complexity in the system, and significant runtime cost, so we decided to remove it.~

~One possibility is to store indices in the attribute, and store the std::strings in a vector elsewhere.~

~I fixed the sentence in the [documentation](https://github.com/BrunoLevy/geogram/wiki/Mesh#attributes)~

**Reply by sebmestrallet** [reply]
Thanks for the answer. Indeed I imagine the induced complexity.

**Reply by BrunoLevy** [reply]
~Yes, it requires quite complicated machinery under the hood. It can be reactivated if there is an absolute need, but I prefer to keep the system as simple as possible (and it is already much more complicated than I wished).~

Finally it was not a big drama, and I already had the basic mechanism (`LifeCycle`) in Graphite, so I moved it to geogram and plugged the wires.
